src/garage/envs/bullet/bullet_env.py

```diff
@@ -25,101 +42,178 @@ def __init__(self, env=None, env_name='', is_image=False):
-            float: Amount of reward returned after previous action
-            bool: Whether the episode has ended, in which case further step()
-                calls will return undefined results
-            dict: Contains auxiliary diagnostic information (helpful for
-                debugging, and sometimes learning)
+            TimeStep: The time step resulting from the action.
+
+        Raises:
+            RuntimeError: if `step()` is called after the environment has been
+                constructed and `reset()` has not been called.
 
         """
+        if self._last_observation is None:
+            raise RuntimeError('reset() must be called before step()!')
+
         observation, reward, done, info = self.env.step(action)
+
+        if self._visualize:
+            self.env.render(mode='human')
+
+        last_obs = self._last_observation
+        # Type conversion
+        if not isinstance(reward, float):
+            reward = float(reward)
+
+        self._last_observation = observation
+        self._step_cnt += 1
+
+        step_type = None
+        if done:
+            step_type = StepType.TERMINAL
+        elif self._step_cnt == 1:
+            step_type = StepType.FIRST
+        else:
+            step_type = StepType.MID
+
         # gym envs that are wrapped in TimeLimit wrapper modify
         # the done/termination signal to be true whenever a time
         # limit expiration occurs. The following statement sets
         # the done signal to be True only if caused by an
         # environment termination, and not a time limit
         # termination. The time limit termination signal
         # will be saved inside env_infos as
-        # 'BulletEnv.TimeLimitTerminated'
-        if 'TimeLimit.truncated' in info:
-            info['BulletEnv.TimeLimitTerminated'] = done  # done = True always
-            done = not info['TimeLimit.truncated']
-        return observation, reward, done, info
+        # 'GarageEnv.TimeLimitTerminated'
+        if 'TimeLimit.truncated' in info or \
+            self._step_cnt >= self._spec.max_episode_length:
+            info['BulletEnv.TimeLimitTerminated'] = True
+            step_type = StepType.TIMEOUT
+        else:
+            info['TimeLimit.truncated'] = False
+            info['BulletEnv.TimeLimitTerminated'] = False
+
+        return TimeStep(
+            env_spec=self.spec,
+            observation=last_obs,
+            action=action,
+            reward=reward,
+            next_observation=observation,
+            env_info=info,
+            agent_info={},  # TODO: can't be populated by env
+            step_type=step_type)
+
+    def render(self, mode):
+        """Renders the environment.
+
+        Args:
+            mode (str): the mode to render with. The string must be present in
+                `self.render_modes`.
+        """
+        if mode not in self.render_modes:
+            raise ValueError('Supported render modes are {}, but '
```

Comment:
if you find that you repeat yourself a lot, you can pull this logic into something like `_validate_render_mode` in `Environment`